src/plugins.c

```diff
@@ -269,66 +259,219 @@ static gint cmp_plugin_names(gconstpointer a, gconstpointer b)
+
+	return PLUGIN_LOADED_OK(p);
+}
 
-	g_module_symbol(plugin->module, "plugin_help", (void *) &plugin->help);
-	g_module_symbol(plugin->module, "plugin_cleanup", (void *) &plugin->cleanup);
-	if (plugin->cleanup == NULL)
+struct LegacyRealHooks
+{
+	void       (*init) (GeanyData *data);
+	GtkWidget* (*configure) (GtkDialog *dialog);
+	void       (*help) (void);
+	void       (*cleanup) (void);
+};
+
+/* Wrappers to support legacy plugins are below */
+static gboolean legacy_init(GeanyPlugin *plugin, gpointer pdata)
+{
+	struct LegacyRealHooks *h = pdata;
+	h->init(plugin->geany_data);
+	return TRUE;
+}
+
+static void legacy_cleanup(GeanyPlugin *plugin, gpointer pdata)
+{
+	struct LegacyRealHooks *h = pdata;
+	/* Can be NULL because it's optional for legacy plugins */
+	if (h->cleanup)
+		h->cleanup();
+}
+
+static void legacy_help(GeanyPlugin *plugin, gpointer pdata)
+{
+	struct LegacyRealHooks *h = pdata;
+	h->help();
+}
+
+static GtkWidget *legacy_configure(GeanyPlugin *plugin, GtkDialog *parent, gpointer pdata)
+{
+	struct LegacyRealHooks *h = pdata;
+	return h->configure(parent);
+}
+
+static void free_legacy_hooks(gpointer data)
+{
+	g_slice_free(struct LegacyRealHooks, data);
+}
+
+/* This function is the equivalent of geany_plugin_register() for legacy-style
+ * plugins which we continue to load for the time being. */
+static void register_legacy_plugin(Plugin *plugin, GModule *module)
+{
+	gint (*p_version_check) (gint abi_version);
+	void (*p_set_info) (PluginInfo *info);
+	void (*p_init) (GeanyData *geany_data);
+	GeanyData **p_geany_data;
+	struct LegacyRealHooks *h;
+
+#define CHECK_HOOK(__x)                                                                   \
+	if (! g_module_symbol(module, "plugin_" #__x, (void *) (&p_##__x)))                   \
+	{                                                                                     \
+		geany_debug("Plugin \"%s\" has no plugin_" #__x "() function - ignoring plugin!", \
+				g_module_name(plugin->module));                                           \
+		return;                                                                           \
+	}
+	CHECK_HOOK(version_check);
+	CHECK_HOOK(set_info);
+	CHECK_HOOK(init);
+#undef CHECK_HOOK
+
+	/* We must verify the version first. If the plugin has become incompatible any
+	 * further actions should be considered invalid and therefore skipped. */
+	if (! plugin_check_version(plugin, p_version_check(GEANY_ABI_VERSION)))
+		return;
+
+	h = g_slice_new(struct LegacyRealHooks);
+
+	/* Since the version check passed we can proceed with setting basic fields and
+	 * calling its set_info() (which might want to call Geany functions already). */
+	g_module_symbol(module, "geany_data", (void *) &p_geany_data);
+	if (p_geany_data)
+		*p_geany_data = &geany_data;
+	/* Read plugin name, etc. name is mandatory but that's enforced in the common code. */
+	plugin->public.info = &plugin->info;
+	plugin->public.geany_data = &geany_data;
```

Comment:
these 2 affectations look redundant, they are done in `plugin_new()` before calling this function